packages/dev/core/src/Materials/meshDebugPluginMaterial.ts

```diff
@@ -0,0 +1,514 @@
+    /**
+     * Whether the mesh debug visualization should multiply with any PBR debug colors underneath.
+     * Defaults to true.
+     */
+    multiplyDebug?: boolean;
+    /**
+     * Default diffuse color of the visualization.
+     * Defaults to (1.0, 1.0, 1.0).
+     */
+    shadedDiffuseColor?: Vector3;
+    /**
+     * Default specular color of the visualization.
+     * Defaults to (0.8, 0.8, 0.8).
+     */
+    shadedSpecularColor?: Vector3;
+    /**
+     * How dark the shading of the visualization is.
+     * Defaults to 10.
+     */
+    shadedSpecularPower?: number;
+    /**
+     * Width of edge lines in TRIANGLES and TRIANGLE_VERTICES modes.
+     * Defaults to 0.7.
+     */
+    wireframeThickness?: number;
+    /**
+     * Color of edge lines in TRIANGLES mode, and color of dots in TRIANGLES_VERTICES and VERTICES mode.
+     * Defaults to (0.0, 0.0, 0.0).
+     */
+    wireframeColor1?: Vector3;
+    /**
+     * Color of edge lines in TRIANGLES_VERTICES modes.
+     * Defaults to (0.8, 0.8, 0.8).
+     */
+    wireframeColor2?: Vector3;
+    /**
+     * Radius of dots drawn over vertices in TRIANGLE_VERTICES and VERTICES mode.
+     * Defaults to 1.2.
+     */
+    vertexRadius?: number;
+    /**
+     * Size of tiles in UV1 or UV2 modes.
+     * Defaults to 20.
+     */
+    uvScale?: number;
+    /**
+     * 1st color of checkerboard grid in UV1 or UV2 modes.
+     * Defaults to (1.0, 1.0, 1.0).
+     */
+    uvColor1?: Vector3;
+    /**
+     * 2nd color of checkerboard grid in UV1 or UV2 modes.
+     * Defaults to (0.5, 0.5, 0.5).
+     */
+    uvColor2?: Vector3;
+    /**
+     * Identifying color of this material in MATERIALIDS mode.
+     * Defaults to a randomly-generated color.
+     */
+    materialColor?: Vector3;
+    /**
+     * Multiplicand for the square root of the camera distance in TRIANGLES_VERTICES and VERTICES mode.
+     * Defaults to .001
+     */
+    distanceScale?: number;
+}
+
+/** @internal */
+export class MeshDebugDefines extends MaterialDefines {
+    DBG_MODE: MeshDebugMode = MeshDebugMode.NONE;
+    DBG_MULTIPLYDEBUG: boolean = true;
+}
+
+/**
+ * Plugin that implements various mesh debug visualizations,
+ * List of available visualizations can be found in MeshDebugMode enum.
+ */
+export class MeshDebugPluginMaterial extends MaterialPluginBase {
+    private _isEnabled: boolean = true;
+    /**
+     * Whether the plugin is enabled or disabled.
+     * Defaults to true.
+     */
+    @serialize()
+    @expandToProperty("_markAllDefinesAsDirty")
+    public isEnabled: boolean = true;
+
+    private _mode: MeshDebugMode = MeshDebugMode.NONE;
+    /**
+     * Current mesh debug mode.
+     * Defaults to NONE.
+     */
+    @serialize()
+    @expandToProperty("_markAllDefinesAsDirty")
+    public mode: MeshDebugMode = MeshDebugMode.NONE;
+
+    /**
+     * Options for the plugin.
+     * See MeshDebugOptions enum for defaults.
+     */
+    private _options: MeshDebugOptions = {};

```

Comment:
There are some TypeScript things you can do here to make this easier to handle. See https://www.typescriptlang.org/docs/handbook/utility-types.html. If you do this, then you won't need all those `!` in the code later.

```suggestion
    private _options: Required<MeshDebugOptions>;
```